src/modules/bruteforce.rs

```diff
@@ -0,0 +1,139 @@
+impl Module for Bruteforce {
+    fn name(&self) -> &'static str {
+        "bruteforce"
+    }
+    fn signature(&self) -> String {
+        "./bruteforce.sh".to_string()
+    }
+    async fn run(&self, app_config: &AppConfig) {
+        let mut rng = rand::thread_rng();
+        let (password, hash_str) = *PASSWORDS_AND_HASHES_LIST.choose(&mut rng).unwrap();
+
+        print(format!("SHA256 value: {hash_str}",)).await;
+        newline().await;
+        csleep(500).await;
+
+        // Wait for "extraction" with a rainbow progress bar
+        {
+            let message = "Extracting Rainbow Table";
+            let width = 30;
+            let millis_wait = 2500;
+            let fill_char = "=";
+
+            let mut spinner_statuses = ["|", "/", "-", "\\"].iter().cycle();
+
+            for i in 0..width {
+                let spinner = if i == (width - 1) {
+                    // Make sure to get a fill_char at the end of our cycle when done.
+                    fill_char
+                } else {
+                    spinner_statuses.next().unwrap()
+                };
+
+                let progress = fill_char.repeat(i);
+                let spaces = " ".repeat(width - i - 1);
+                let progress_content = rainbow(&format!("{progress}{spinner}{spaces}"));
+
+                print(format!("\r{message} [{progress_content}]",)).await;
+
+                csleep((millis_wait / width) as u64).await;
+
+                if app_config.should_exit() {
+                    return;
+                }
+            }
+
+            newline().await;
+        }
+
+        print("Begin matching").await;
+        newline().await;
+        csleep(500).await;
+
+        // Show the progress of "decryption"
+        {
+            let mut progress: usize = 0;
+
+            let hash_bytes = hash_str.bytes().collect::<Vec<_>>();
+            let l = hash_bytes.len();
+
+            while progress < l {
+                let guesses: Vec<_> = gen_hex_string(&mut rng, l as u64).bytes().collect();
+
+                while progress < l && guesses[progress] == hash_bytes[progress] {
+                    progress += 1;
+                }
+
+                let (done, undone) = if progress < l {
+                    let done = from_utf8(&hash_bytes[0..progress]).unwrap();
+                    let undone = from_utf8(&guesses[progress..]).unwrap();
+                    (done, undone)
+                } else {
+                    (hash_str, "")
+                };
+
+                let (done, undone) = (Paint::green(done), Paint::red(undone));
+
+                print(format!("\r :: {done}{undone} ::")).await;
+                csleep(10).await;
+
+                if app_config.should_exit() {
+                    return;
+                }
+            }
+            newline().await;
+        }
+
+        print(format!("+ Match found -- the password is \"{password}\"")).await;
+        newline().await;
+    }
+}
+
+// color a string rainbow
```

Comment:
Might it be more maintainable to use [colorful](https://docs.rs/colorful/latest/colorful/trait.Colorful.html) here? We will have to check whether it breaks the web version, though.

Reply:
I didn't know that crates!
I'll look into it.

Reply:
Can we use 24-bit colors?
`colorful` seems not to work without 24-bit color support.

Even in 2022 there are a few terminal emulators without the support (macOS terminal, for example)

Reply:
By the way, can I use 256 colors instead of ANSI 8 colors?
It will make the step of the rainbow colors smoother.

Reply:
I found [colorgrad](https://github.com/mazznoer/colorgrad-rs), which generates gradient colors.
I reimplemented the rainbow progress bar with it, along with the smoother colors using 256 colors in 098bfd1.

![Screenshot 2022-10-08 at 12 45 23](https://user-images.githubusercontent.com/8597717/194686609-fcf7fec8-bde0-464d-ab1c-77ef26ebbe17.png)